File: apscparser.py
import urllib2, os
from BeautifulSoup import *

def apscparser(sem):
	#create subfolder called fall or winter
	try: 
		os.makedirs(sem)
	except:
		pass
	#grab the URL remotely and start parsing it
	url = "http://www.apsc.utoronto.ca/timetable/"+sem+".html"
	site=urllib2.urlopen(url)
	soup=BeautifulSoup(site.read())
	#Examine all HTML tables in the site
	for table in soup.findAll('table'):
		mytable = []
		rows = table.findAll('tr')
		#convert to 2D array first
		for i in range (0,len(rows)):
			cols = rows[i].findAll('td')
			try:
				new = []
				for j in range(0,7):
					#replace all &nbsps with blanks - easier to check
					if cols[j].find(text=True)=="&nbsp;":
						new.append("")
					else:
						new.append(cols[j].find(text=True))
				mytable.append(new)
			except:
				pass
		#then convert to csv
		for i in range (0,len(mytable)):
			row = mytable[i]
			#if the section is blank, then we want to copy the value from above
			if row[1]=="" and row[0]==mytable[i-1][0] and row[2]>mytable[i-1][2]:
				for j in range(i-1,-1,-1):
					#but only if the meet (row[2]) is different from the one above
					#sometimes the same section will be taking place in multiple
					#locations at the same time and we only want to grab one of
					#those rows
					if mytable[j][0]==row[0] and mytable[j][1]!="" and mytable[j][2]<row[2]:
						row[1] = mytable[j][1]
						break
			#strip out info like notes from the location cell
			row[6] = row[6].split('*')[0]
			row[6] = row[6].split(' ')[0]
			row[6] = row[6].split('-')[0]
			row[6] = row[6].split('\r')[0]
			row[6] = row[6].split('\n')[0]
			#get department code, e.g. it is CSC if the course is CSC180H1F
			dept = row[0][0:3]
			#ignore anything marked blank or PEY or with a blank section
			if len(dept)>1 and dept != "PEY" and row[1] != "":
				#open (or create if not found) a csv file to store the timings
				#filenames are formatted like ECE_fall.csv, CSC_winter.csv, etc
				#they are stored in either the "fall" or "winter" subfolders
				f = open(sem+"/"+dept+'_'+sem+".csv",'a')
				#don't need to show the Meet value
				line = ",".join(row[j] for j in [0,1,3,4,5,6])
				f.write(line+'\n')
				f.close()

File: apscparser.pl
# This program crawls the Engineering timetable pages from U of T
# It extracts the listings and places them into .csv files. 
use strict;
use warnings;
use LWP::Simple;
use HTML::TableExtract;
use Text::CSV;

#this function would be called with either "fall" or "winter" as arguments
#It extracts the tables from the HTML pages and converts them to csv
sub apscparse {
	#semester = the only function argument
	my $sem = $_[0];
	#open fall.html or winter.html
	my $raw_html = do {
		open my $in, '<', "$sem.html"
		    or die "Can't open infile: $!\n";
		local $/ = undef;
		<$in>;
	};

	#We only want these five columns from the page. Details like location and prof
	#are less important. The main focus is timing.
	my $te = new HTML::TableExtract(headers => [qw(NAME SECTION MEET DAY START FINISH LOCATION)]);
	#this creates an array of tables, each table corresponding to one department
	#the tables are stored in memory first before being saved to csv
	$te->parse($raw_html);

	#Examine all resulting tables
	foreach my $ts ($te->tables) {
		#to avoid reference errors, copy the table into this local variable
		my @table = $ts->rows;
		#count number of rows in the table
		my $numrows = $#table+1;
		#loop through the table and analyze it
		for (my $i = 0; $i<$numrows; $i++) {
			my $name = $table[$i][0];
			my $sect = $table[$i][1];
			#strip out info like notes from the location cell
			$table[$i][6] =~ s/\ .*//g;
			$table[$i][6] =~ s/\r.*//g;
			$table[$i][6] =~ s/\n.*//g;
			$table[$i][6] =~ s/-.*//g;
			$table[$i][6] =~ s/\*.*//g;
			#if the section is blank, then we want to copy the value from above
			#would prefer to check if it's null/undef/zero
			#but those comparisons always seem to fail
			if (length($sect)<5){
				#check if the course code is identical to the row above
				if ($name eq $table[$i-1][0]){
					#if so, copy the section (LEC 01, PRA 02, etc), from the
					#above row into the current row.
					$table[$i][1] = $table[$i-1][1];
				}
			}
			#extract the department 3-letter code, e.g. ECE, CSC, MIE, etc
			my $dept = $table[0][0];
			$dept = substr $dept,0,3;
			#ignore the "course" marked PEY500Y1Y (why is it even there...)
			if ($dept ne "PEY"){
				#open (or create if not found) a csv file to store the timings
				#filenames are formatted like ECE_fall.csv, CSC_winter.csv, etc
				#they are stored in either the "fall" or "winter" subfolders
				open (FH, ">>$sem/$dept"."_$sem.csv") or die "$!";
				if ($i==0 || !(($table[$i][0] eq $table[$i-1][0]) && 
						($table[$i][1] eq $table[$i-1][1]) &&
						($table[$i][2] eq $table[$i-1][2]))){
					my @row = @{$table[$i]};
					splice @row,2,1;
					#output each row into the csv file
					print FH join( ',', @row ), "\n";
				}
			}
		}
	}
}

#call that major function from above twice
apscparse("fall");
apscparse("winter");
